### PR TITLE
fix(sampler): give each sampler its own PRNG stream (#1245)

### LINKED ISSUE #1245: mlx_lm.server: seed parameter silently ignored
## `mlx_lm.server`: `seed` parameter silently ignored

Chat-completions endpoint accepts `seed` and routes it through `mx.random.seed`, but output is the same across different seed values for the same prompt. mlx-lm 0.31.3, M-series, reproduced on Qwen2.5 and Llama-3.1 4 bit quants.

## Repro

Start server:

```bash
MODEL_PATH=/path/to/your/mlx/model
mlx_lm.server --model "$MODEL_PATH" --port 1234
```

Send three requests with different seeds (same shell, `$MODEL_PATH` carries over):

```bash
for s in 11111 22222 33333; do
  curl -sS http://127.0.0.1:1234/v1/chat/completions \
    -H 'Content-Type: application/json' \
    -d "{\"model\":\"$MODEL_PATH\",\"temperature\":1,\"max_tokens\":25,\"seed\":$s,\"stream\":false,\"messages\":[{\"role\":\"user\",\"content\":\"Tell me a one-sentence story about a cat.\"}]}" \
  | jq -r '.choices[0].message.content'
done
```
```
bash repro.sh
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
The tabby cat's reflection in the dark window was the only companion she hadn't driven away that evening.
```

All three responses come back byte-identical.

## Expected

Different seeds should follow different sampling trajectories which should provide different outputs (at temperature > 0).

## My thoughts why

`server.py:1191` reads `seed` from the body into `self.seed`. At `server.py:956-957` it's applied:

```python
if args.seed is not None:
    mx.random.seed(args.seed)
```

That call executes. Sampling lives at `sample_utils.py:267-282`:

```python
@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)
def categorical_sampling(logits, temp):
    return mx.random.categorical(logits * (1 / temp))
```

~Hypothesis: `mx.compile` with `inputs=mx.random.state` captures the state object at decoration time, so subsequent `mx.random.seed(...)` calls don't propagate into what the compiled function reads. Effectively, sampling becomes deterministic against whatever state existed at module import.~
### Update

Initial suspicion was that the `mx.compile` decorator captured a stale `mx.random.state` reference. Not actually it: running `mx.random.seed(N) -> mx.random.categorical(...)` directly with three different seeds produces three different draws, both with and without the decorator. So it looks like the bug is somewhere downstream of the seed call but upstream of (or alongside) the actual sampling.
## Also tried

- Removed `seed` from the body, ran the same prompt twice. Identical output. State isn't naturally advancing across requests either.
- Same prompt at temperature 0.5 / 1.5 / 3.0. Outputs differ. Sampling runs, temperature is honored. Seed/state path is broken.
- Different prompts at fixed seed. Outputs differ. Not response-caching.

## Workaround

Vary something in the prompt per request (a hidden tag in the system message works). At least breaks the unintended determinism.

### PR DESCRIPTION
**Pull request by sashko-zakharchuk**
Fixes #1245.

## The bug

`/v1/chat/completions` accepts and validates a `seed`, but it has no effect on the output, and identical requests return byte-identical tokens at temperature > 0.

Measured on `MiniMax-M3-MLX-6bit`, 5 identical requests per cell at `temperature=1.0, top_p=0.95`:

```
noseed: distinct=1/5   max_repeat=[359,359,359,359,359]
seeded: distinct=1/5   (seeds 1..5 all returning the same text)
```

#1245 also reports it on Qwen2.5 and Llama-3.1.

## Cause

Every sampling draw reads the implicit global `mx.random.state`, and the sampling functions are `mx.compile`d with `inputs=mx.random.state`. The issue thread suspected the compile decorator captures the state object, and the reporter's own testing did not confirm that — so I have deliberately not built the fix on that diagnosis. Threading explicit keys removes the dependency on global state entirely, so it holds regardless of the precise mechanism.

Worth noting separately: a global RNG cannot give per-request reproducibility on a server at all. `BatchGenerator` runs multiple generations concurrently, so with one shared stream a request's output depends on what else is in flight.

## The change

`make_sampler` takes an optional `seed` and owns an explicit key chain, splitting a subkey per randomness consumer — each sampling method, plus the final categorical draw. `apply_xtc` and `categorical_sampling` take an optional `key`. `server.py` derives fresh entropy when a request carries no seed, and no longer calls `mx.random.seed` globally, which was resetting the RNG for any concurrently generating request.

A sampler is already constructed per request in `server.py` and per sequence in `BatchGenerator`, so per-sampler state gives per-request isolation with:

- **no changes to `generate.py`** and no arity change at any of the four sampler call sites
- **no public API break** — the documented `Callable[[mx.array], mx.array]` signature still works, so third-party custom samplers are unaffected
- **compilation preserved** — the key is an array input, so nothing is un-compiled

With no explicit seed, one is derived from the global RNG rather than leaving the key as `None`. That keeps each sampler independent, keeps a caller's `mx.random.seed(...)` honoured, and avoids passing `None` into the compiled functions on the hot path.

## Result

Same measurement after the change — 4/4 cells:

```
noseed: distinct=5/5   degen=0/5
seeded: distinct=5/5   degen=0/5
```

## Tests

`tests/test_sampler_rng.py` covers: a seed reproduces the stream; different seeds diverge; interleaved samplers do not perturb each other; the full `top_p`/`min_p`/`top_k` chain stays reproducible; the XTC path is seeded; `temp==0` stays argmax; the legacy one-argument call still works; and the compiled helpers still accept a call with no key.

Verified end to end on Apple Silicon (M3 Ultra, macOS 26.5, mlx 0.32.0) against a live `mlx_lm.server`.

### CODE REVIEW

**Comment by sashko-zakharchuk** [comment]
Closing this in favour of a corrected patch.

I re-reviewed it before anyone spent time on it, and found defects I'd rather fix than have reviewed:

- **Distributed serving would diverge.** `_make_sampler` seeded each request with `secrets.randbits(32)` — per-process entropy, so every rank derives a different seed and ranks stop emitting identical tokens. The server synchronises `mx.random.state` across ranks precisely to avoid that.
- **The tests need rebuilding.** I'm not satisfied they fail against unpatched `main`, which would make them worthless as regression tests.
- Smaller: an existing in-repo test breaks, `black` fails on the changed files, and an out-of-range client seed can kill the generation thread.

The underlying bug in #1245 is real and reproducible — identical requests return byte-identical tokens at temperature > 0 (measured: 1 distinct output across 5 identical requests), and the `seed` value has no effect. A revised, much smaller PR will follow.
